File: application/src/main/java/de/pbauerochse/worklogviewer/fx/tasks/CheckForUpdateTask.java
package de.pbauerochse.worklogviewer.fx.tasks;

import com.fasterxml.jackson.core.type.TypeReference;
import de.pbauerochse.worklogviewer.tasks.Progress;
import de.pbauerochse.worklogviewer.util.ExceptionUtil;
import de.pbauerochse.worklogviewer.util.HttpClientUtil;
import de.pbauerochse.worklogviewer.util.JacksonUtil;
import de.pbauerochse.worklogviewer.version.GitHubVersion;
import org.apache.http.client.methods.CloseableHttpResponse;
import org.apache.http.client.methods.HttpGet;
import org.apache.http.impl.client.CloseableHttpClient;
import org.apache.http.util.EntityUtils;
import org.jetbrains.annotations.NotNull;
import org.jetbrains.annotations.Nullable;
import org.slf4j.Logger;
import org.slf4j.LoggerFactory;

import java.io.IOException;
import java.io.StringReader;
import java.util.Collections;
import java.util.List;
import java.util.Optional;

import static de.pbauerochse.worklogviewer.util.FormattingUtil.getFormatted;
import static java.util.Comparator.comparing;

/**
 * Fetches the most recent version of the
 * Worklog Viewer from Github
 */
public class CheckForUpdateTask extends WorklogViewerTask<Optional<GitHubVersion>> {

    private static final Logger LOGGER = LoggerFactory.getLogger(CheckForUpdateTask.class);

    public CheckForUpdateTask() {
        super(getFormatted("task.updatecheck"));
    }

    @Override
    @Nullable
    public Optional<GitHubVersion> start(@NotNull Progress progress) {
        progress.setProgress(getFormatted("worker.updatecheck.checking"), 0.1);
        Optional<GitHubVersion> versionOptional = getVersionFromGithub();

        progress.setProgress(getFormatted("worker.progress.done"), 100);
        return versionOptional;
    }

    private Optional<GitHubVersion> getVersionFromGithub() {
        return getAllGithubReleaseVersions().stream()
                .filter(GitHubVersion::isRelease)
                .max(comparing(GitHubVersion::getPublished));
    }

    private List<GitHubVersion> getAllGithubReleaseVersions() {
        HttpGet request = new HttpGet("https://api.github.com/repos/pbauerochse/youtrack-worklog-viewer/releases");
        request.addHeader("Accept", "application/json, text/plain, */*");

        try (CloseableHttpClient client = getHttpClient()) {
            try (CloseableHttpResponse response = client.execute(request)) {
                if (HttpClientUtil.isValidResponseCode(response.getStatusLine())) {
                    String jsonResponse = EntityUtils.toString(response.getEntity());
                    return JacksonUtil.parseValue(new StringReader(jsonResponse), new TypeReference<>() {
                    });
                }

                LOGGER.warn("Could not get most recent version: {}", response.getStatusLine().getReasonPhrase());
                EntityUtils.consumeQuietly(response.getEntity());
            }
        } catch (IOException e) {
            throw ExceptionUtil.getIllegalStateException("exceptions.updater.versions", e);
        }

        return Collections.emptyList();
    }

    private CloseableHttpClient getHttpClient() {
        return HttpClientUtil.getDefaultClientBuilder(1)
                .setDefaultHeaders(HttpClientUtil.getRegularBrowserHeaders())
                .build();
    }
}

File: application/src/main/java/de/pbauerochse/worklogviewer/fx/tasks/CheckForUpdateTask.kt
package de.pbauerochse.worklogviewer.fx.tasks

import com.fasterxml.jackson.core.type.TypeReference
import com.fasterxml.jackson.module.kotlin.jacksonObjectMapper
import de.pbauerochse.worklogviewer.tasks.Progress
import de.pbauerochse.worklogviewer.tasks.WorklogViewerTask
import de.pbauerochse.worklogviewer.util.ExceptionUtil
import de.pbauerochse.worklogviewer.util.FormattingUtil.getFormatted
import de.pbauerochse.worklogviewer.version.GitHubVersion
import org.apache.hc.client5.http.classic.methods.HttpGet
import org.apache.hc.client5.http.config.ConnectionConfig
import org.apache.hc.client5.http.config.RequestConfig
import org.apache.hc.client5.http.impl.classic.CloseableHttpClient
import org.apache.hc.client5.http.impl.classic.HttpClients
import org.apache.hc.client5.http.impl.io.BasicHttpClientConnectionManager
import org.apache.hc.client5.http.impl.routing.SystemDefaultRoutePlanner
import org.apache.hc.core5.http.Header
import org.apache.hc.core5.http.HttpStatus
import org.apache.hc.core5.http.io.entity.EntityUtils
import org.apache.hc.core5.http.message.BasicHeader
import org.apache.hc.core5.http.message.StatusLine
import org.slf4j.LoggerFactory
import java.io.IOException
import java.net.ProxySelector
import java.util.Comparator.comparing
import java.util.concurrent.TimeUnit

/**
 * Fetches the most recent version of the
 * Worklog Viewer from Github
 */
class CheckForUpdateTask : WorklogViewerTask<GitHubVersion?>(getFormatted("task.updatecheck")) {

    override fun start(progress: Progress): GitHubVersion? {
        progress.setProgress(getFormatted("worker.updatecheck.checking"), 0.1)
        val version = mostRecentGithubReleaseVersion

        progress.setProgress(getFormatted("worker.progress.done"), 100)
        return version
    }

    private val mostRecentGithubReleaseVersion: GitHubVersion?
        get() = allGithubReleaseVersions.stream()
            .filter { it.isRelease }
            .max(comparing { it.published })
            .orElse(null)

    private val allGithubReleaseVersions: List<GitHubVersion>
        get() {
            val request = HttpGet("https://api.github.com/repos/pbauerochse/youtrack-worklog-viewer/releases")
            request.addHeader("Accept", "application/json, text/plain, */*")

            try {
                httpClient.use { client ->
                    client.execute(request) { response ->
                        val statusLine = StatusLine(response)
                        if (isValidResponseCode(statusLine)) {
                            return@execute MAPPER.readValue(response.entity.content, object : TypeReference<List<GitHubVersion>>() {})
                        }

                        LOGGER.warn("Could not get most recent version: {}", statusLine.reasonPhrase)
                        return@execute EntityUtils.consumeQuietly(response.entity)
                    }
                }
            } catch (e: IOException) {
                throw ExceptionUtil.getIllegalStateException("exceptions.updater.versions", e)
            }

            return emptyList()
        }

    private val httpClient: CloseableHttpClient
        get() {
            val connectionConfig = ConnectionConfig.custom()
                .setConnectTimeout(REQUEST_TIMEOUT_IN_MILLIS, TimeUnit.MILLISECONDS)
                .build()


            val config = RequestConfig
                .custom()
                .setConnectionRequestTimeout(REQUEST_TIMEOUT_IN_MILLIS, TimeUnit.MILLISECONDS)
                .build()

            return HttpClients.custom()
                .setDefaultRequestConfig(config)
                .setConnectionManager(BasicHttpClientConnectionManager().apply { setConnectionConfig(connectionConfig) })
                .setRoutePlanner(SystemDefaultRoutePlanner(ProxySelector.getDefault()))
                .setDefaultHeaders(WEB_BROWSER_HEADERS)
                .build()
        }

    private fun isValidResponseCode(statusLine: StatusLine?): Boolean {
        if (statusLine == null) throw ExceptionUtil.getIllegalArgumentException("exceptions.main.worker.nullstatus")
        val statusCode = statusLine.statusCode
        return statusCode >= HttpStatus.SC_OK && statusCode < HttpStatus.SC_MULTIPLE_CHOICES
    }

    companion object {
        private val LOGGER = LoggerFactory.getLogger(CheckForUpdateTask::class.java)
        private val MAPPER = jacksonObjectMapper()
        private const val REQUEST_TIMEOUT_IN_MILLIS = 2000L
        private val WEB_BROWSER_HEADERS: List<Header> = listOf(
            BasicHeader("User-Agent", "Mozilla/5.0 (X11; Linux x86_64) AppleWebKit/537.36 (KHTML, like Gecko) Chrome/41.0.2272.118 Safari/537.36"),
            BasicHeader("Accept-Encoding", "gzip, deflate, sdch"),
            BasicHeader("Accept-Language", "de-DE,de;q=0.8,en-US;q=0.6,en;q=0.4")
        )
    }
}
